rust/rsmgp-sys/src/value/mod.rs

```diff
@@ -0,0 +1,370 @@
+use crate::map::*;
+use crate::memgraph::*;
+use crate::mgp::*;
+use crate::path::*;
+use crate::result::*;
+use crate::vertex::*;
+// Required here, if not present tests linking fails.
+#[double]
+use crate::mgp::ffi;
+use mockall_double::double;
+
+/// # Safety
+/// TODO(gitbuda): Write section about safety.
+pub(crate) unsafe fn create_cstring(c_char_ptr: *const i8) -> MgpResult<CString> {
+    match CString::new(CStr::from_ptr(c_char_ptr).to_bytes()) {
+        Ok(v) => Ok(v),
+        Err(_) => Err(MgpError::UnableToMakeCString),
+    }
+}
+
+// NOTE on why mutable pointer to mgp_value has to be owned by this code.
+//
+// mgp_value used to return data is non-const, owned by the module code. Function to delete
+// mgp_value is non-const.
+// mgp_value containing data from Memgraph is const.
+//
+// `make` functions return non-const value that has to be deleted.
+// `get_property` functions return non-const copied value that has to be deleted.
+// `mgp_property` holds *const mgp_value.
+//
+// Possible solutions:
+//   * An enum containing *mut and *const can work but the implementation would also contain
+//   duplicated code.
+//   * A generic data type seems complex to implement https://stackoverflow.com/questions/40317860.
+//   * Holding a *const mgp_value + an ownership flag + convert to *mut when delete function has to
+//   be called.
+//   * Hold only *mut mgp_value... as soon as there is *const mgp_value, make a copy (own *mut
+//   mgp_value). The same applies for all other data types, e.g. mgp_edge, mgp_vertex.
+//   mgp_value_make_vertex accepts *mut mgp_vartex (required to return user data), but data from
+//   the graph is all *const T.
+//
+// The decision is to move on with a copy of mgp_value because it's cheap to make the copy.
+
+/// Useful to own `mgp_value` coming from / going into Memgraph as a result.
+///
+/// Underlying pointer object is going to be automatically deleted.
+#[derive(Debug)]
+pub struct MgpValue {
+    // It's not wise to create a new MgpValue out of the existing value pointer because drop with a
+    // valid pointer will be called multiple times -> double free problem.
+    ptr: *mut mgp_value,
+}
+
+impl Drop for MgpValue {
+    fn drop(&mut self) {
+        unsafe {
+            if !self.ptr.is_null() {
+                ffi::mgp_value_destroy(self.ptr);
+            }
+        }
+    }
+}
+
+impl MgpValue {
+    pub fn new(ptr: *mut mgp_value) -> MgpValue {
+        MgpValue { ptr }
+    }
+
+    pub(crate) fn mgp_ptr(&self) -> *const mgp_value {
+        self.ptr
+    }
+
+    pub fn to_value(&self, memgraph: &Memgraph) -> MgpResult<Value> {
+        unsafe { mgp_raw_value_to_value(self.mgp_ptr(), &memgraph) }
+    }
+
+    pub fn make_null(memgraph: &Memgraph) -> MgpResult<MgpValue> {
+        unsafe {
+            let mgp_ptr = ffi::mgp_value_make_null(memgraph.memory());
+            if mgp_ptr.is_null() {
+                return Err(MgpError::UnableToMakeNullValue);
+            }
```

Comment:
It is funny that we return an error if the result is null when we are making a null value... :smile: